### ISSUE TITLE
False positive on switch 'no duplicate case'

### ISSUE THREAD

**Issue by kentaromiura** [eslint/eslint#3408]
In EcmaScript after the `case` token there can be any valid expression,
in particular this switch was raising a false `no duplicate case` positive:
```
switch(['a','b'].toString()){
  case (['a','b']).toString():
    console.log('right');
    break
  case (['a']).toString():
    console.log('not right');
    break
}    
```

in this case is because the [getHash function](https://github.com/eslint/eslint/blob/master/lib/rules/no-duplicate-case.js#L21-L35) does not handle ArrayExpression nodes;
the [if else cascade](https://github.com/eslint/eslint/blob/master/lib/rules/no-duplicate-case.js#L22-L34) exit without matching, so the function returns `undefined`.
This happens for every node types that isn't checked.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!
